text/0030-protocol-messages.md

```diff
@@ -0,0 +1,246 @@
+being used by the node and is used to match the connection with a potential white-listed peer.
+
+<sup>2</sup> Bit-masks are used to denote what protocol versions the node supports. The LSB acts as a starting point.
+Up to 32 bytes are supported, limiting the number of protocol versions to 256. Examples:
+* `[0b00000001]` denotes that the node supports protocol version 1.
+* `[0b00000111]` denotes that the node supports protocol versions 1, 2 and 3.
+* `[0b01101110]` denotes that the node supports protocol versions 2, 3, 4, 6 and 7.
+*  `[0b01101110, 0b01010001]` denotes that the node supports protocol versions 2, 3, 4, 6, 7, 9, 13 and 15.
+* `[0b01101110, 0b01010001, 0b00010001]` denotes that the node supports protocol versions 2, 3, 4, 6, 7, 9, 13, 15, 17
+  and 21.
+
+### Version 1
+
+#### `LegacyGossip`
+
+Type ID: `2`
+
+A legacy message to broadcast a transaction and request another one at the same time.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`transaction`|Transaction to broadcast. Can be compressed<sup>1</sup>.|`Vec<u8>`|292-1604|
+|`hash`|Hash of the requested transaction.|`[u8; 49]`|49|
+
+<sup>1</sup> Compression is detailed at the end.
+
+**Note**: This message is the original IRI protocol message before the TLV protocol was introduced. It was kept by
+HORNET for compatibility with IRI but is not used between HORNET nodes. Its "ping-pong" concept has complex consequences
+on the node design and as such will not be implemented by Bee.
+
+### Version 2
+
+#### `MilestoneRequest`
+
+Type ID: `3`
+
+A message to request a milestone.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`index`|Index of the requested milestone.|`u32`|4|
+
+#### `TransactionBroadcast`
+
+Type ID: `4`
+
+A message to broadcast a transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`transaction`|Transaction to broadcast. Can be compressed<sup>1</sup>.|`Vec<u8>`|292-1604|
+
+<sup>1</sup> Compression is detailed at the end.
+
+#### `TransactionRequest`
+
+Type ID: `5`
+
+A message to request a transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`hash`|Hash of the requested transaction.|`[u8; 49]`|49|
+
+#### `Heartbeat`
+
+Type ID: `6`
+
+A message that informs about the part of the tangle currently being fully stored by a node.
+This message is sent when a node:
+* just got paired to another node;
+* did a local snapshot and pruned away a part of the tangle;
+* solidified a new milestone;
+
+It also helps other nodes to know if they can ask it a specific transaction.
+
+|Name|Description|Type|Length|
+|----|-----------|----|------|
+|`solid_milestone_index`|Index of the last solid milestone.|`u32`|4|
+|`snapshot_milestone_index`|Index of the snapshotted milestone.|`u32`|4|
+
+### Compression
+
+A transaction encoded in bytes has a length of `1604`. The `payload` field itself occupies `1312` bytes and is often
```

Comment:
maybe too nitpicky again, but I would mention the T5B1 encoding here, something a long the lines of:
"A transaction encoded using the T5B1 codec has a length of `1604`.